include/kernels/staggered_prolong_restrict.cuh

```diff
@@ -0,0 +1,142 @@
+  }
+
+  // Function to return the coarse ColorSpinorField
+  template<StaggeredTransferType transferType>
+  inline const ColorSpinorField& coarseColorSpinorField(const ColorSpinorField& quoteIn, const ColorSpinorField&)
+  {
+    errorQuda("Invalid transfer type %d for coarseColorSpinorField", (int)transferType);
+    return quoteIn;
+  }
+
+  // on prolong, the out vector is the fine vector
+  template<>
+  inline const ColorSpinorField& coarseColorSpinorField<StaggeredTransferType::STAGGERED_TRANSFER_PROLONG>(const ColorSpinorField& quoteIn, const ColorSpinorField&) {
+    return quoteIn;
+  }
+
+  // on restrict, the in vector is the fine vector
+  template<>
+  inline const ColorSpinorField& coarseColorSpinorField<StaggeredTransferType::STAGGERED_TRANSFER_RESTRICT>(const ColorSpinorField&, const ColorSpinorField& quoteOut) {
+    return quoteOut;
+  }
+
+  /**
+      Kernel argument struct
+  */
+  template <typename Float, int fineSpin, int fineColor_, int coarseSpin, int coarseColor, QudaFieldOrder order, StaggeredTransferType theTransferType>
+  struct StaggeredProlongRestrictArg : kernel_param<> {
+    static constexpr int fineColor = fineColor_;
+    FieldOrderCB<Float, StaggeredTransferOutSpin<fineSpin,coarseSpin,theTransferType>::outSpin, StaggeredTransferOutColor<fineColor,coarseColor,theTransferType>::outColor,1,order> out;
+    const FieldOrderCB<Float, StaggeredTransferInSpin<fineSpin,coarseSpin,theTransferType>::inSpin, StaggeredTransferInColor<fineColor,coarseColor,theTransferType>::inColor,1,order> in;
+    const int *geo_map;  // need to make a device copy of this
+    const spin_mapper<fineSpin,coarseSpin> spin_map;
+    const int parity; // the parity of the output field (if single parity)
+    const int nParity; // number of parities of input fine field
+    const int fineX[4]; // fine spatial volume
+    const int fineVolumeCB; // fine spatial volume
+    const int coarseVolumeCB; // coarse spatial volume
+    static constexpr StaggeredTransferType transferType = theTransferType;
+
+    StaggeredProlongRestrictArg(ColorSpinorField &out, const ColorSpinorField &in,
+                                const int *geo_map, const int parity) :
+      out(out), in(in), geo_map(geo_map), spin_map(), parity(parity),
+      nParity(fineColorSpinorField<transferType>(in,out).SiteSubset()),
+      fineX{fineColorSpinorField<transferType>(in,out).X()[0],
+        fineColorSpinorField<transferType>(in,out).X()[1],
+        fineColorSpinorField<transferType>(in,out).X()[2],
+        fineColorSpinorField<transferType>(in,out).X()[3]},
+      fineVolumeCB(fineColorSpinorField<transferType>(in,out).VolumeCB()),
+      coarseVolumeCB(coarseColorSpinorField<transferType>(in,out).VolumeCB())
+    {
+      this->threads = dim3(fineVolumeCB, nParity, fineColor);
+    }
+  };
+
+  /**
+     Performs the permutation from a coarse degree of freedom to a
+     fine degree of freedom
+  */
+  template <typename Arg> struct StaggeredProlongRestrict_ {
```

Comment:
No underscore at end of `StaggeredProlongRestrict_`

Reply:
83ec85e5ffb9bccc66ad43ed832d912433d1aa52